src/main.rs

```diff
@@ -42,15 +49,139 @@ fn read_can(mut publisher: MqttClient, can_interface: &str) {
+    client.connect().expect("Could not connect to Siren!");
+    let rx = client.start_consumer().expect("Could not begin consuming");
+    client
+        .subscribe("Calypso/Bidir/Command/#")
+        .expect("Could not subscribe!");
+
+    // do the default initialization for all, do outside of the thread to save time negotiating when send_can comes up
+    let mut writable_send_map = send_map.write().expect("Could not modify send messages!");
+    for key in ENCODABLE_KEY_LIST {
+        let packet = EncodableMessage::new(String::clone(&key.to_owned()), Vec::new());
+        let ret = packet.encode();
+        writable_send_map.insert(ret.id, ret);
     }
+    drop(writable_send_map);
+
+    let join_handle = thread::spawn(move || {
+        for msg in rx.iter() {
+            if let Some(msg) = msg {
+                let buf = match command_data::CommandData::parse_from_bytes(msg.payload()) {
+                    Ok(buf) => buf,
+                    Err(err) => {
+                        println!("Could not decode command: {}", err);
+                        continue;
+                    }
+                };
+                let key = match msg.topic().split('/').last() {
+                    Some(key) => key.to_owned(),
+                    None => {
+                        println!("Could not parse the key value in {}", msg.topic());
+                        continue;
+                    }
+                };
+
+                let packet = EncodableMessage::new(String::clone(&key), buf.data);
+                let ret = packet.encode();
+                send_map
+                    .write()
+                    .expect("Could not modify send messages!")
+                    .insert(ret.id, ret);
+            } else {
+                // the code doesnt work without this else statement
+                // idk why but never remove this else statement
+                let is_conn = client.is_connected();
+                println!("Client is {}", is_conn);
+                if !is_conn {
+                    println!("Trying to reconnect");
+                    match client.reconnect() {
+                        Ok(_) => println!("Reconnected!"),
+                        Err(_) => println!("Could not reconnect!"),
+                    }
+                    continue;
+                }
+            }
+        }
+    });
+    Ok(join_handle)
+}
+
+fn send_out(
+    can_interface: &str,
+    send_map: Arc<RwLock<HashMap<u32, EncodeData>>>,
+) -> Result<JoinHandle<()>, String> {
+    let socket = CanSocket::open(can_interface).expect("Failed to open CAN socket!");
+
+    let join_handle = thread::spawn(move || loop {
+        thread::sleep(Duration::from_millis(750));
+        let sender = send_map.read().expect("Cannot read map of sendables!");
+        for msg in sender.iter() {
+            // let id = u32::from_str_radix((msg.1.1).trim_start_matches("0x"), 16).expect("Invalid CAN ID!");
```

Comment:
purpose of this?

Reply:
the comment was how to get the id as a u32 from a string, but I decided to move that to the python layer and forgot to remove it.